vinty.8
.TH "VINTY" "8"
.SH "NAME"
vinty \-\- manage OpenSSL certificates
.SH "SYNOPSIS"
.B vinty \-\-init

.B vinty \-\-create
client|server
common-name

.B vinty \-\-renew
client|server
common-name

.B vinty \-\-revoke
common-name

.B vinty \-\-list
valid|revoked|all
.SH "DESCRIPTION"
This manual page documents
.BR vinty .
.B vinty
is used to manage (create, revoke, renew) OpenSSL certificate/key pairs.
It can only create self-signed certificates, as its primary use is
creating certificate/key pairs for use with OpenVPN.
.PP
.B vinty
is an alternative to the PKI web application. Using both at the same time
is not recommended, since certificate/key pairs created by the web application
are owned by the user running the webserver. Ceritificates created by
.B vinty
cannot be read by the web application.
.SH OPTIONS
.TP
.B \-\-init
Initialise the PKI infrastructure. This will create a Certificate Authority,
generate the necessary Diffie Hellman parameters and create an initial CRL
(Certificate Revocation List).
.TP
.B \-\-create
Create a certificate/key pair of a given type (either
.B client
or
.B server\fR\|)
with the given
.BR common-name .
.TP
.B \-\-renew
Renew a certificate/key pair of a given type (either
.B client
or
.B server\fR\|)
and with a given
.BR common-name .
.TP
.B \-\-revoke
Revoke a certificate of a given
.BR common-name .
Additionally, the CRL will be regenerated.
.TP
.B \-\-list
Print out a list of certificates issued by the Certificate Authority.
The type of certificates must be specified. You can choose to show
only valid, revoked or all certificates.
.TP
.B \-\-help, \-h
Show a help message and exit.
.TP
.B \-\-version
Show version info and exit.
.SH "FILES"
.TP
.B /etc/vinty/vinty.cfg
The main configuration file of
.BR vinty .
It contains the necessary information to be embedded in the certificates,
the key size, and specifies where the keys are stored.
.TP
.B /etc/vinty/openssl.cnf
The OpenSSL configuration file used by
.BR vinty .
.PP
The web application uses 2 additional configuration files:
.TP
.B /etc/vinty/apache.conf
The necessary Apache configuration to use the
.BR vinty
web frontend.
.TP
.B /etc/vinty/scripts.conf
Configuration file sourced by external scripts.
.SH "EXAMPLES"
Create a client certificate with
.B test1.example.com
as common-name:

.B vinty \-\-create client test1.example.com

Renew the certificate with common-name
.B test1.example.com
(which is a client certificate):

.B vinty \-\-renew client test1.example.com

Revoke the certificate with common-name
.B test1.example.com

.B vinty \-\-revoke test1.example.com

Show all certificates issued by the Certificate Authority:

.B vinty \-\-list all
.SH "BUGS"
When using the web application, all certificates and keys will be owned by
the user running the webserver.
.PP
When using the commandline tool, certificates and keys are owned by the
.B root
user.
.PP
Certificates created with the commandline tool are thus not accessible
from the web application, as they are owned by
.BR root .
.SH "SEE ALSO"
\fIopenssl\fR\|(1), \fIvintypasswd\fR\|(1), \fIopenvpn\fR\|(8)


debian/changelog
vinty (0.1) unstable; urgency=low

  * Prepare for initial release

 -- Jeroen Hooyberghs <jeroen.hooyberghs@fks.be>  Fri, 07 Oct 2011 10:39:40 +0100


vinty.py
#!/usr/bin/python

## Revision history ############################################################
## Revision history ############################################################
__author__  = 'Wouter Eerdekens <info@fks.be>'
__date__    = '2011-08-12'
__version__ = '0.1'
__history__ = """
2011-08-12 - Prepare for initial release <jeroen.hooyberghs@fks.be>
2006-07-26 - initial version.
"""
## Imports #####################################################################
import os, sys
from os import path
from optparse import OptionParser
import pki.cmd, pki.util

## Functions ###################################################################
def error(msg):
   sys.stderr.write('Error: ' + msg + '\n')
   sys.exit(1)

# ------------------------------------------------------------------------------
def parse_argv(argv):
   cn      = ''
   usage   = """%prog --init
       %prog --create client|server CN
       %prog --renew client|server CN
       %prog --revoke CN
       %prog --list valid|revoked|all"""
   version = '%prog ' +  __version__
   parser  = OptionParser(usage=usage, version=version)
   parser.add_option('--init', action='store_true', default=False,
                     help='initialise the PKI infrastructure')
   parser.add_option('--list', metavar='TYPE', type='choice',
                     choices=('valid', 'revoked', 'all'),
                     help='list all certificates of TYPE')
   parser.add_option('--create', metavar='TYPE', type='choice',
                     choices=(pki.CERT_CLIENT, pki.CERT_SERVER),
                     help='create a certificate/key pair of TYPE')
   parser.add_option('--renew', metavar='TYPE', type='choice',
                     choices=(pki.CERT_CLIENT, pki.CERT_SERVER),
                     help='renew a certificate of TYPE')
   parser.add_option('--revoke', action='store_true', default=False,
                     help='revoke a certificate')
   opt, arg = parser.parse_args(argv)
   if opt.init:
      if opt.create or opt.renew or opt.revoke or opt.list:
         parser.error('--init cannot be combined with other options')
      if len(arg) != 0:
         parser.error('--init doesn\'t take an argument')
   elif opt.list:
      if opt.init or opt.create or opt.renew or opt.revoke:
         parser.error('--list cannot be combined with other options')
      if len(arg) != 0:
         parser.error('--list doesn\'t take extra arguments')
   elif opt.create:
      if opt.init or opt.list or opt.renew or opt.revoke:
         parser.error('--create cannot be combined with other options')
      if len(arg) == 1:
         cn = arg[0]
      else:
         parser.error('--create requires you to specify a CN')
   elif opt.renew:
      if opt.init or opt.create or opt.revoke or opt.list:
         parser.error('--renew cannot be combined with other options')
      if len(arg) == 1:
         cn = arg[0]
      else:
         parser.error('--renew requires you to specify a CN')
   elif opt.revoke:
      if opt.init or opt.create or opt.renew or opt.list:
         parser.error('--revoke cannot be combined with other options')
      if len(arg) == 1:
         cn = arg[0]
      else:
         parser.error('--revoke requires you to specify a CN')
   else:
      parser.print_help()
      sys.exit(1)
   return opt, cn

# ------------------------------------------------------------------------------
def info(msg):
   sys.stdout.write(msg)
   sys.stdout.flush()

# ------------------------------------------------------------------------------
def pki_init(config):
   sys.stdout.write('Creating PKI infrastructure: ')
   sys.stdout.flush()
   result, msg = pki.cmd.create_pki(config)
   if result:
      print 'ok'
   else:
      print 'failed'
      error(msg)

# ------------------------------------------------------------------------------
def cert_create(config, cn):
   type = opt.create
   cn   = pki.util.strip_invalid(cn)
   if cn:
      info('Creating ' + type + ' certificate with CN ' + cn + ': ')
      result, msg = pki.cmd.create_cert(config, cn, type)
      if result:
         print 'ok (serial: ' + pki.util.serial_from_cn(config, cn) + ')'
      else:
         print 'failed'
         error(msg)
   else:
      error('invalid CN given')

# ------------------------------------------------------------------------------
def cert_renew(config, cn):
   type = opt.renew
   cn   = pki.util.strip_invalid(cn)
   if cn:
      serial = pki.util.serial_from_cn(config, cn)
      if serial:
         info('Renewing ' + type + ' certificate with CN ' + cn + ': ')
         result, msg = pki.cmd.renew_cert(config, cn, opt.renew)
         if result:
            print ' ok (new serial: ' + pki.util.serial_from_cn(config,cn) + ')'
         else:
            print ' failed'
            error(msg)
      else:
         error('no serial found for ' + cn)
   else:
      error('invalid CN given')

# ------------------------------------------------------------------------------
def cert_revoke(config, cn):
   cn = pki.util.strip_invalid(cn)
   if cn:
      serial = pki.util.serial_from_cn(config, cn)
      if serial:
         info('Revoking certificate with CN ' + cn + ', serial ' + serial +': ')
         result, msg = pki.cmd.revoke_cert(config, cn, opt.revoke)
         if result:
            print ' ok'
         else:
            print ' failed'
            error(msg)
      else:
         error('no serial found for ' + cn)
   else:
      error('invalid CN given')

## We're being called as a script ##############################################
if __name__ == '__main__':
   opt, cn = parse_argv(sys.argv[1:])
   config  = pki.util.parse_config()
   if not config:
      error('unable to read/validate configuration')
   if opt.init:
      pki_init(config)
   else:
      if not pki.cmd.check_pki(config):
         error('PKI infrastructure not valid (first run with --init)')
      if opt.list:
         pki.cmd.list_certs(config, opt.list)
      elif opt.create:
         cert_create(config, cn)
      elif opt.renew:
         cert_renew(config, cn)
      elif opt.revoke:
         cert_revoke(config, cn)
